### DISCUSSION TITLE
Tuple + Array

### DISCUSSION THREAD

**Discussion by StefanTerdell** [colinhacks/zod#3443] [Q&A] [answered]
Hey friends!

I'm looking for a way to replicate the following type:
```typescript
type X = [string, boolean, ...number[]]
```

In Json Schema this would be something like this (in +2020 replace `"items"` with `"prefixItems"`):
```json
{
  "type": "array",
  "items": [
    {
      "type": "string"
    },
    {
      "type": "number"
    }
  ],
  "additionalItems": {
    "type": "number"
  },
  "minItems": 2
}
```

Any ideas? V4 candidate perhaps?

**Comment by chrissikora** [comment, marked as answer]
Try `z.tuple([ z.string(), z.boolean()]).rest(z.number())`

**Comment by tamerkalla** [comment]
Worth checking if you convert schemas for tool calling: Zod 4.4.3 emits {"type":"array","prefixItems":[{"type":"string"},{"type":"number"}]} for z.tuple([z.string(), z.number()]) — no minItems, no items: false. So the JSON Schema the model sees accepts [], ["a"] and ["a",1,2], all of which Zod then rejects at runtime, after the provider already approved the call. ArkType and TypeBox both emit the length constraints for the same shape. I found it with schema-parity.
